src/java/com/google/tagpost/spanner/SpannerService.java

```diff
@@ -0,0 +1,90 @@
+package com.google.tagpost.spanner;
+
+import com.google.cloud.Timestamp;
+import com.google.cloud.spanner.*;
+import com.google.tagpost.Comment;
+import com.google.tagpost.Tag;
+import com.google.tagpost.Thread;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/** Data access and operations of Spanner */
+public class SpannerService implements DataService {
+
+  @Override
+  public List<Thread> getAllThreadsByTag(Tag tag) {
+
+    SpannerOptions spannerOptions = SpannerOptions.newBuilder().build();
+    Spanner spanner = spannerOptions.getService();
+    List<Thread> threadList = new ArrayList<>();
+
+    try {
+      DatabaseId db = DatabaseId.of("testing-bigtest", "tagpost", "test");
+      DatabaseClient dbClient = spanner.getDatabaseClient(db);
+      String query =
+          "SELECT ThreadID, PrimaryTag FROM Thread WHERE PrimaryTag = '" + tag.getTagName() + "'";
+      ResultSet resultSet = dbClient.singleUse().executeQuery(Statement.of(query));
+      convertResultToThreadList(threadList, resultSet);
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+    return threadList;
+  }
+
+  @Override
+  public List<Comment> getAllCommentsByThreadId(long threadId) {
+
+    SpannerOptions spannerOptions = SpannerOptions.newBuilder().build();
+    Spanner spanner = spannerOptions.getService();
+    List<Comment> commentList = new ArrayList<>();
+
+    try {
+      DatabaseId db = DatabaseId.of("testing-bigtest", "tagpost", "test");
+      DatabaseClient dbClient = spanner.getDatabaseClient(db);
+      String query = "SELECT * FROM Comment WHERE ThreadID = " + threadId;
+      ResultSet resultSet = dbClient.singleUse().executeQuery(Statement.of(query));
+      convertResultToCommentList(threadId, commentList, resultSet);
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+    return commentList;
+  }
+
+  private void convertResultToThreadList(List<Thread> threadList, ResultSet resultSet) {
```

Comment:
1. Move `List<Thread>` to the return value.
1. Prefer immutable types like `ImmutableList` as return values, if you don't have special needs.

Reply:
Fixed in latest commit.